### PR TITLE
Dashboard to have human-friendly message with grouping

### LINKED ISSUE #390: Better description of actual linkage error on component page
```
log4j-1.2.12.jar (4 errors):
  ClassSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.CommunicatorServer, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  ClassSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.HtmlAdaptorServer, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.HtmlAdaptorServer, methodName=<init>, interfaceMethod=false, descriptor=()V}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.CommunicatorServer, methodName=start, interfaceMethod=false, descriptor=()V}, reason: CLASS_NOT_FOUND, target class location not found
```

e.g.

```
org.apache.log4j.jmx.Agent references com.sun.jdmk.comm.HtmlAdaptorServer which is not found. 
org.apache.log4j.jmx.Agent references com.sun.jdmk.comm.HtmlAdaptorServe.start which is not found. 
```

### LINKED ISSUE #391: When reporting static linkage errors notice if they're all the same problem
E.g. here it's all about the same class not found

```
log4j-1.2.12.jar (4 errors):
  ClassSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.CommunicatorServer, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  ClassSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.HtmlAdaptorServer, subclass=false}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.HtmlAdaptorServer, methodName=<init>, interfaceMethod=false, descriptor=()V}, reason: CLASS_NOT_FOUND, target class location not found
  MethodSymbolReference{sourceClassName=org.apache.log4j.jmx.Agent, targetClassName=com.sun.jdmk.comm.CommunicatorServer, methodName=start, interfaceMethod=false, descriptor=()V}, reason: CLASS_NOT_FOUND, target class location not found
```

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #390 : introducing human-friendly message (`JarLinkageReport.formatByGroup` using `LinkageErrorCause.toString()`)
Fixes #391 by grouping errors by newly added `LinkageErrorCause` (symbol name + reason)

image

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 1]
```diff
@@ -78,6 +86,55 @@ public String toString() {
...
+    ImmutableListMultimap<LinkageErrorCause, StaticLinkageError<ClassSymbolReference>>
+        groupedClassErrors = Multimaps.index(getMissingClassErrors(), LinkageErrorCause::from);
+
+    ImmutableListMultimap<LinkageErrorCause, StaticLinkageError<MethodSymbolReference>>
+        groupedMethodErrors = Multimaps.index(getMissingMethodErrors(), LinkageErrorCause::from);
+
+    ImmutableListMultimap<LinkageErrorCause, StaticLinkageError<FieldSymbolReference>>
+        groupedFieldErrors = Multimaps.index(getMissingFieldErrors(), LinkageErrorCause::from);
+
+    // ImmutableSet ensures deterministic iteration order
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: It's strange to expect reproducible order for any set. Would a list work?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+package com.google.cloud.tools.opensource.classpath;
+
+import com.google.auto.value.AutoValue;
+
+/** Key to group static linkage errors by their causes. */
+@AutoValue
+abstract class LinkageErrorCause {
+
+  /** Returns the reason of error */
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: of --> for the

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    StringBuilder builder = new StringBuilder(getSymbol());
+    switch (getReason()) {
+      case CLASS_NOT_FOUND:
+        builder.append(" is not found. ");
+        break;
+      case INACCESSIBLE_CLASS:
+        builder.append(" is not accessible. ");
+        break;
+      case INCOMPATIBLE_CLASS_CHANGE:
+        builder.append(" has changed incompatible. ");
```
Comment (elharo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: changed incompatibly

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/JarLinkageReportTest.java [hunk 4]
```diff
@@ -74,44 +87,56 @@ public void setUp() {
...
             + "fieldName=fieldX}, reason: CLASS_NOT_FOUND, target class location not found\n",
         jarLinkageReport.toString());
   }
 
+  @Test
+  public void testFormatByGroup() {
+    Assert.assertEquals(
+        "3 group(s) of 4 static linkage error(s)\n"
+            + "  ClassA is not found. Referenced by: ClassB, ClassC\n"
+            + "  ClassA.methodX is not found in the class. Referenced by: ClassB\n"
```
Comment (elharo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/JarLinkageReportTest.java]: probably don't need "in the class" here

Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/JarLinkageReportTest.java]: Removed.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java (hunk 3 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: Updated.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java (hunk 2 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: Fixed

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: No, lists don't work for this case; `combinedKeys` needs to remove duplicates. 

ImmutableSet guarantees iteration order https://github.com/google/guava/wiki/ImmutableCollectionsExplained#how

**Pull request merged**